### ISSUE TITLE
Strange cache lag

### ISSUE THREAD

**Issue by nextfullstorm** [MapWindow/MapWinGIS#182]
I'm experiencing some odd behavior and want to clear things up. So I'm using mwg from delphi and recently started to dig in mwg sources to speed up tiles loading. I'm using local filesystem as a source for tiles (btw, support for local filesystem reverted from master, is it intentionally? Changelog mentions local fs support in 5.1). First load of every tile brings noticeable lag, you can see how tiles draws one by one even on very fast CPU. It was expected behavior for me - we need to load png, decode it to bmp, draw on canvas, store to db, store to ram cache etc. That's how I was thinking... until I found a way to hack it around. Let's see:

1. Delete mwtiles db: load is slow because of cache adding cost (omit png/jpg decoding)?

2. Delete mwtiles AND disable sqlite cache via _diskCache.useCache = false, _diskCache.doCaching = false in TileManager::InitCaches(): slow because of... because of cache absence? No cached tiles so we need to load every tile from local fs which is slow (SSD, btw)?

3. Delete mwtiles, enable sqlite cache AND patch SQLiteCache.cpp:368 with following code:
```diff
--- "a/src/Tiles/Caching/SQLiteCache.cpp"
+++ "b/src/Tiles/Caching/SQLiteCache.cpp"
@@ -367,6 +367,25 @@ bool SQLiteCache::get_TileExists(BaseProvider* provider, int scale, int x, int y
 // Extracts a tile from the database
 TileCore* SQLiteCache::get_Tile(BaseProvider* provider, int scale, int x, int y)
 {
+	try
+	{
+		for (size_t i = 0; i < provider->get_SubProviders()->size(); i++)
+		{
+			auto fmt = (*provider->get_SubProviders())[i]->get_UrlFormat();
+			if (fmt.Find("file:///", 0) == 0)
+			{
+				return provider->GetTileImage(CPoint(x, y), scale);
+			}
+		}
+	}
+	catch (const std::exception&)
+	{
+		CallbackHelper::ErrorMsg("SQLiteCache: exception on getting tile.");
+		return nullptr;
+	}
+	
 	TileCore* tile = NULL;
 	
 	if(!Initialize(SqliteOpenMode::OpenIfExists))
```
What do we do here? We replace sqlite cache retrieving with direct fs retrieving. When mwg tries to load next tile from sqlite, it instead loads tile directly from file. And it loads blazing fast! 

So to sum it up: local fs tile load is slow with empty sqlite cache and with disabled cache, but very fast with "pass-through" cache.

I found this behavior very odd. If there is no overhead in loading directly from fs, why do we see overhead if we disable sqlite caching?

**Comment by jerryfaust** [comment]
Hello @nextfullstorm 

I'm not well-versed on Tiles, but I can try to step through and compare the execution paths.
1. Where do you get your local tile files.  Is this something you download from OSM perhaps?
2. Regarding the local file system 'reverted' from 'master', that was my doing.  Your pull request was against 'master', and I went ahead and accepted it, but then realized that I shouldn't have.  All updates should do against 'develop', and only get integrated into 'master' at the next release.  So I had to revert that update.  That said, it looks like it was not then integrated into 'develop'.  So that is something I can look at.

**Comment by nextfullstorm** [comment]
Hello @jerryfaust,

1. You're right, i'm using OSM tiles, you can get my test dataset (about 46mb) from [here](https://yadi.sk/d/KgOg9jqCDCg1qg), use 47.2, 39.6 as a start geocoords. You need to re-apply #152 to get it work. Unpack anywhere, create custom provider and set url like `file:///c:\temp\maps\{zoom}\{x}\{y}.png`. Feel free to contact me if anything goes wrong.

**Comment by jerryfaust** [comment]
Initial results:
1. disabled your patch that bypasses cache

```
TileCore* SQLiteCache::get_Tile(BaseProvider* provider, int scale, int x, int y)
{
	//try
	//{
	//	for (size_t i = 0; i < provider->get_SubProviders()->size(); i++)
	//	{
	//		auto fmt = (*provider->get_SubProviders())[i]->get_UrlFormat();
	//		if (fmt.Find("file:///", 0) == 0)
	//		{
	//			return provider->GetTileImage(CPoint(x, y), scale);
	//		}
	//	}
	//}
	//catch (const std::exception&)
	//{
	//	CallbackHelper::ErrorMsg("SQLiteCache: exception on getting tile.");
	//	return nullptr;
	//}

	TileCore* tile = NULL;

```
2. turned off all caching internally

```
void TileManager::InitCaches()
{
    _ramCache.doCaching = false; // true;
    _ramCache.useCache = false; // true;
    _ramCache.cache = TileCacheManager::get_Cache(tctRamCache);
    _caches.push_back(&_ramCache);

    _diskCache.doCaching = false; // m_globalSettings.wmsDiskCaching;
    _diskCache.useCache = false; // true;
    _diskCache.cache = TileCacheManager::get_Cache(tctSqliteCache);
    _caches.push_back(&_diskCache);
}

```

Unless I'm overlooking something, I am not seeing the lagging.  Could something else be going on?

[Here is a link to a video of the test app](https://jerryfaust1-gmail.tinytake.com/tt/MzkwMzMxN18xMTk2Mzg5MQ).

**Comment by nextfullstorm** [comment]
> I am not seeing the lagging

Me neither. In your video, at least. But things different on my side.

1. Disabled SQLiteCache::get_Tile hack. Disabled caches. Like in your 1,2. Result - https://drive.google.com/open?id=1GVuRM6tzsSRRQazwsn3pE7TE47kYqIU8.

2. Enabled hack, enabled caching, like in my OP. Result - https://drive.google.com/open?id=18wtSpjxx1uZYPyC1Bk-qBKLv2hgXdp2U.

**Comment by jerryfaust** [comment]
I have a couple thoughts moving forward:
1. If you want to send me a simple exe that I could run through the debugger to see what calls are made and perhaps what may be causing the lag.  This is probably the easiest and best place to start.
2. I have the Community edition of Delphi, if perhaps you could help me get set up with simple project.  I don't think Delphi is the issue, since you can load fast by simply bypassing the cache.  This option just allows me to work from the same base-point as you.

What do you think?

**Comment by nextfullstorm** [comment]
Why not both? Here https://drive.google.com/open?id=12LC3INWHk71dd_LpqxgnzOT7B7a9MLSZ you can find test project and compiled exe (don't forget to check it on virustotal.com, safety first).

How to use test app: press "..." and select folder with map, then press "start". No checks, no error handling.

What if you want to build it yourself. Well, if you don't have much experience with delphi, this might be a problem (at least for me it was). Basically you need to build component project and then install component TmwMap into delphi IDE. You can download this project here - https://drive.google.com/open?id=1uK9QeUJs9vrhqYSFe13612F2UxYj5hWv, or you can google any tutorial on how to install activex component into delphi and follow it (remember to replace *TLB.pas with my edition). Right click on project in Project Manager, press Build, then press Install. If everything went ok and you see "package installed" with long list of installed packages, then **RESTART** IDE and you probably ready to build test app. If anything goes wrong, well, it's delphi. Comment here then, we'll work it out somehow. And I did not test any of these in Community version.

**Comment by jerryfaust** [comment]
I've attached links to 3 videos.

First running your app with No Cache, but enabling your hack so that it essentially bypasses OCX code to directly load the images.
[DelphiAppNoCacheWithHack](https://drive.google.com/file/d/1bYG4hm500dtJdPgrmTtR3HCdthUmS23O/view?usp=sharing)

Then disabling the hack so that it goes through the OCX code to load the images, with no caching.
[DelphiAppNoCacheNoHack](https://drive.google.com/file/d/1V9jIyBrLnFTbNJ3FMoL3Fz2utWTRDKeT/view?usp=sharing)

Finally, with the same build of the ocx (no hack, no cache), switching over to my basic .NET app.
[NetAppNoCacheNoHack](https://drive.google.com/file/d/1tA1qP9Yqe0O7rtpKylEX012DLhwPo5Rh/view?usp=sharing)

I'm not sure what the interpretation is.  They're all very similar.  The first two are essentially the same.  The .Net app _may be_ slightly faster, but only negligibly.  They all, however, seem faster than your earlier submission.  I notice that this sample app you sent is not the same as in your video (which appeared to load more slowly).  Comparing apples to apples, how does this simple app, running on your system, compare with what I'm seeing?

FYI, I'm running on a virtual Win7 machine on an SSD (albeit running through the Visual Studio debugger, so they could all benefit from an unhindered release build).

Regards.

**Comment by nextfullstorm** [comment]
You did not test most interesting combination - **with** cache enabled **and** hack. No cache means there is no requests to SQLiteCache::get_Tile. That's why you didn't see any difference - hack not involved. You need to enable caching and insert sqlite hack.

Regarding apples to apples. Simple app with cache disabled - https://drive.google.com/open?id=1f1NX8iBeGY63Oecd0ZkkkhX3eg1-JA7V. Simple app with cache enabled and hack enabled - https://drive.google.com/open?id=1JBCYuQILN64emGQtS6ZotzLZaDeKFf6z. I ran it on Win7 VM as well to match your test environment. For better (I mean - slower) results you should maximize window - it works significantly slower when there is more tiles to load (my screen size is 1920*1080).

And regarding .Net app - I think you should disable zoom animation to make fair comparison. 

Cheers.

**Comment by jerryfaust** [comment]
Yes, of course. You're exactly right.  And yes, I now see the blazing difference.

So I will need more time to revisit this, but on the first run-through, here's what I think is happening.  Your 'patch' short-circuits the cache, and loads every tile from disk, thinking that it is finding it in the cache.  And so it never executes any code beyond that point; it thinks it found every tile in the cache.

What would have happened if it did not find some of the tiles in the cache, is to set up a thread-pool of 'tasks' to load each of the required tiles.  So my initial thought is that there is more overhead to set up the thread-pool to load the tiles, than there is to simply load all tiles from disk on the single thread.

Perhaps it makes more sense to set up the thread pool for web-based tiles, since there is more latency involved.  But for local disk access, it's simply faster to fetch them all in-line.  Another thought (since even our web-based tile loading is not very speedy) is to try omitting the thread-pool even for web-based tiles, just to compare performance.

We recently switched to using the cURL library for web-based tiles.  But we are only using the synchronous 'easy' library, not the asynchronous 'multi' library.  This is something that I have also considered revisiting.  If someone out there is very familiar with the cURL library, feel free to review our implementation and recommend updates.

Thanks again for pointing this out.  I'll let you know what I learn.

**Comment by jerryfaust** [comment]
Ok, I was able to continue on this.  Here are my current findings.
1. I'm convinced that the thread-pool introduces enough overhead to make file-based tiles load more slowly than loading one-after-another on a single thread.  HOWEVER, this may only be on SSD-based systems.  It may be that spindle-based disks introduce enough latency that the thread-pool is indeed beneficial.  This will require more testing on different system configurations.
2. I am likewise convinced that the thread-pool definitely benefits http-based tile loading.  Stepping through the code, I found the problem with our current implementation.  There was a critical-section lock entering the GetTileHttpData method.  This was forcing threads to enter single-file, effectively loading only one http tile at a time.  By removing this lock, the tile loading is improved significantly.  We will need to watch for any behavioral issues, in case any of the code paths are not thread-safe; but in a cursory overview, they do appear to be (e.g. not relying on any shared variables).  I increased the thread pool size from the default of 5 up to 10, then 15, with increasing benefit to the load time.

I will be creating a JIRA ticket and submitting a couple updates.  One to re-introduce file-based tile support, and another to remove the critical section.  Regarding the file-based tiles, I am tempted to include your patch so that file-based tiling bypasses the cache altogether, but I think rather that I will wait on that change until we can show that it benefits all system types (but I'm open to any input).

Regards,
Jerry.

**Comment by nextfullstorm** [comment]
I did quick research and can't confirm thread-pool overhead you mentioned. But I don't know codebase good enough and don't have time to dig deeper at the moment. As I can see, pool creation executes at first load and after that it's no-op. Maybe request execution brings this overhead, I don't have experience with CThreadPool at all.

As for my 'patch': I don't think things should work this way. It's obviously a hack. More correct way would be:

1. Eliminate thread pool overhead, thus we can simply disable caching via documented API and get fast load times.

-or-

2. Extract file-based tile provider in separate entity with separate codepath without thread pools, caches and so on.

**Comment by jerryfaust** [comment]
I should note here that two updates have been submitted that are related to this issue:
1. [MWGIS-205: GdiplusStartup/Shutdown should not be called in DllMain](https://mapwindow.atlassian.net/browse/MWGIS-205), and
2. [MWGIS-207: Allow multi-threaded HTTP Tile loading](https://mapwindow.atlassian.net/browse/MWGIS-207)

A third update is forthcoming, that of re-introducing file-based tiles from the local file system.

**Comment by pmeems** [comment]
If this is still an issue with the latest release (https://github.com/MapWindow/MapWinGIS/releases) could you create a new issue?
For now, I close this old issue.
